plugins/database/cassandra/connection_producer_test.go

```diff
@@ -0,0 +1,109 @@
+package cassandra
+
+import (
+	"context"
+	"crypto/tls"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/gocql/gocql"
+	"github.com/hashicorp/vault/helper/testhelpers/cassandra"
+	"github.com/hashicorp/vault/sdk/database/dbplugin/v5"
+	"github.com/stretchr/testify/require"
+)
+
+var (
+	insecureFileMounts = map[string]string{
+		"test-fixtures/no_tls/cassandra.yaml": "/etc/cassandra/cassandra.yaml",
+	}
+	secureFileMounts = map[string]string{
+		"test-fixtures/with_tls/cassandra.yaml": "/etc/cassandra/cassandra.yaml",
+		"test-fixtures/with_tls/keystore.jks":   "/etc/cassandra/keystore.jks",
+		"test-fixtures/with_tls/.cassandra":     "/root/.cassandra/",
+	}
+)
+
+func TestTLSConnection(t *testing.T) {
+	type testCase struct {
+		config    map[string]interface{}
+		expectErr bool
+		// errorMsg is only checked if expectErr is true. This also a partial string match, so if this value shows up
+		// anywhere in the error it will pass the assertion
+		errorMsg string
+	}
+
+	tests := map[string]testCase{
+		"tls not specified": {
+			config:    map[string]interface{}{},
+			expectErr: true,
+			errorMsg:  "EOF",
+		},
+		"unrecognized certificate": {
+			config: map[string]interface{}{
+				"tls": "true",
+			},
+			expectErr: true,
+			errorMsg:  "certificate signed by unknown authority",
+		},
+		"insecure TLS": {
+			config: map[string]interface{}{
+				"tls":          "true",
+				"insecure_tls": true,
+			},
+			expectErr: false,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			cleanup, connURL := cassandra.PrepareTestContainer(t,
+				cassandra.Version("3.11.9"),
+				cassandra.CopyFromTo(secureFileMounts),
+				cassandra.SslOpts(&gocql.SslOptions{
+					Config:                 &tls.Config{InsecureSkipVerify: true},
+					EnableHostVerification: false,
+				}),
+			)
+			defer cleanup()
+
+			// This is fine if it panics - the connection URL should include the port and if it doesn't, this should fail
+			port := strings.Split(connURL, ":")[1]
```

Comment:
Should we handle this failure a bit more gracefully by doing a length check first instead of letting it panic?

Reply:
I don't think an additional check is valuable here because it's in a test and it needs to fail, so having it be a test failure rather than a panic doesn't really give us much. It's an irrecoverable error due to the test construct failing which seems like a panic is well suited for.

Reply:
Further investigation revealed that panicking in a test does not behave the same as running `t.Fatal`. It will stop all tests that are running in the package. I didn't realize this was the behavior. I've updated this so the `PrepareTestContainer` function instead returns a new `Host` type that contains the hostname, the port, and has a function to build the connection URL. It also performs the equivalent check in `PrepareTestContainer` rather than forcing the callers to make any additional checks.